nxos_gir timeout fixes

chrisvanheuveln opened the pull request:
##### SUMMARY
`nxos_gir` is getting a timeout-related failure when using `httpapi` transport.

This PR supercedes #64907 

Analysis: The CLI command `system mode maintenence` often requires at least 120 seconds to complete. The standard timeout is 30 seconds but users can override this timeout by setting environment variable `ANSIBLE_PERSISTENT_COMMAND_TIMEOUT`; however, this env var was ignored when the `httpapi` codepath was used. 

Further, since this extended minimum timeout is specific to `nxos_gir`, we felt it would be beneficial to automatically set the timeout for this module.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`nxos_gir`

##### ADDITIONAL INFORMATION
Tested on `N9K, N9Kv, N6K, N7K, N3K` (internal testbeds: `dt-n9k5-1, greensboro-nx-1, hamilton-nx-1, n6k-77, n7k-99, n3k-105`

chrisvanheuveln committed:
nxos_gir timeout fixes

chrisvanheuveln commented on lib/ansible/module_utils/network/nxos/nxos.py:
```diff
@@ -635,6 +636,18 @@ def edit_config(self, candidate=None, commit=True, replace=None, comment=None):
+                conn.set_option('persistent_command_timeout', gir_timeout)
+                msg = "timeout value extended to %ss for nxos_gir." % gir_timeout
+                self._module.warn(msg)
+
```
Note that this method exists in the `httpapi` class so it doesn't need to check for other transport states.

ganeshrn replied:
The logic to set command timeout can be moved to nxos action plugin
```
        conn = Connection(self._connection.socket_path)
        timeout = conn.get_option('persistent_command_timeout')
        conn.set_options(direct={'persistent_command_timeout': 241})
```

chrisvanheuveln committed:
additional setup/teardown cleanups for sanity test

NilashishC reviewed (commented):
Is there any reason why were are not handling this the way we did for nxos_install_os in the nxos action plugin?

mikewiebe commented:
> Is there any reason why were are not handling this the way we did for nxos_install_os in the nxos action plugin?

As we discussed earlier today, this module is more predictable then `nxos_install_os` so we feel it's a nice usability enhancement to set the default timeout automatically instead of forcing the user to set the timeout manually.

chrisvanheuveln commented:
> Is there any reason why were are not handling this the way we did for nxos_install_os in the nxos action plugin?

@NilashishC The problem as I see it is that the `Connection` object is different in the action plugin. Any settings changed there are not inherited when the httpapi plugin is called to send commands to the device.

As a test I removed the new gir code; then changed the action plugin so that it set the timeout value to 31.

This is the general call sequence when I run a playbook. I don't understand the mechanics enough to explain why httpapi plugin `__init__` is called twice, but notice that the action plugin uses the first `Connection` object, while the httpapi plugin uses the second object.

```
-> plugins/httpapi/nxos.py:
     class HttpApi: __init__()
       self.connection == <ansible.plugins.connection.httpapi.Connection object at 0x10a0c2450>

-> plugins/httpapi/nxos.py:
     class HttpApi: __init__()
       self.connection == <ansible.plugins.connection.httpapi.Connection object at 0x105027390>

-> plugins/action/nxos.py:
     class ActionModule: run()
       self._connection == <ansible.plugins.connection.httpapi.Connection object at 0x10a0c2450>
       self._connection.set_options(direct={'persistent_command_timeout': 31})   <--------<< timeout now changed **
       self._connection.get_option('persistent_command_timeout') == 31

-> network/nxos/nxos_gir:
     main()

-> plugins/httpapi/nxos.py:
     class HttpApi: _run_queue()
       self.connection == <ansible.plugins.connection.httpapi.Connection object at 0x105027390>
       self.connection.get_option('persistent_command_timeout') == 30    <---<< different Connection object has default value **

```

chrisvanheuveln committed:
Add timeout extension comment to module header

chrisvanheuveln commented:
@NilashishC I added the timeout comment to the module as requested.

chrisvanheuveln committed:
lint

NilashishC commented:
@chrisvanheuveln It seems that adding the following in the action plugin overrides the command_timeout value for network_cli. Could you try this out?
```
conn = Connection(self._connection.socket_path)
conn.set_options(direct={'persistent_command_timeout', 240})
```

chrisvanheuveln commented:
@NilashishC @ganeshrn 
This is the code I added to the action plugin:
```
                from ansible.module_utils.connection import Connection
                conn = Connection(self._connection.socket_path)
                timeout = conn.get_option('persistent_command_timeout')
                conn.set_options(direct={'persistent_command_timeout', 241})
```

`get_option` works while `set_options` fails with this error:

```
*** ConnectionError: Failed to encode some variables as JSON for communication with ansible-connection. 
The original exception was: set([241, 'persistent_command_timeout']) is not JSON serializable
```
Mike mentioned this might be due to python versioning, as I'm using python `2.7.15`

ganeshrn commented:
@chrisvanheuveln The comment from @NilashishC has a typo. It should be `conn.set_options(direct={'persistent_command_timeout': 241})` instead of `conn.set_options(direct={'persistent_command_timeout', 241})`

chrisvanheuveln commented:
@ganeshrn Your codefix isn't working. The call to set_options appears to work but a side effect is that it seems to have corrupted / overwrote the session handle in the process. I added my diffset/playbook/findings to a gist here: https://gist.github.com/chrisvanheuveln/fe4364ef9d7c9f804fd218c0b7e64210

chrisvanheuveln committed:
rework logic to use action plugin

chrisvanheuveln commented:
@ganeshrn Thanks for the help. Your suggestion to use `set_option` instead of `set_options` works fine from the action plugin so I removed the changes from `module_utils`.

chrisvanheuveln commented:
recheck

chrisvanheuveln commented:
recheck

chrisvanheuveln commented:
@NilashishC All requested changes are complete. Any more comments or can we merge this PR? Thanks

NilashishC commented:
@chrisvanheuveln Thank you for making all the changes and apologize for the delay in getting this merged.
